### DISCUSSION TITLE
var / let / const と言う脈略のない三銃士ども

### DISCUSSION THREAD

**Discussion by ubugeeei** [Ubugeeei/work-log#136] [Scraps]
# 序
ここ数日、for文の由来は何か? と言う話題が上がった。
シンタックスのキーワードの由来については自分も関心があって、好きなトピック。
中でも、いつも思っていたことがあって、JavaScriptの変数宣言に使われるキーワード選定が謎と言うのがある。
このことについてツイートしたところたくさんの方に反応をいただけたので備忘録として。

元ツイ: https://twitter.com/ubugeeei/status/1691491304428236800?s=20
お気持ち: https://twitter.com/ubugeeei/status/1691623230812123376?s=20

その他お気持ち:
https://twitter.com/ubugeeei/status/1691519021957672961?s=20
https://twitter.com/ubugeeei/status/1691523247718952963?s=20
~~https://twitter.com/ubugeeei/status/1691527941300318212?s=20~~

var については歴史的背景が伺える。まだ let も const もなかった時代だ。
まあ変数を宣言するのに var と言うキーワードを持ち寄るのは自然なことだろう。

# 議題
問題は let と const の由来。
ともにブロックスコープをとると言うのは共通として、let は再代入が可能、constは再代入が不可能 (immutableなわけではない)と言うもの。
百歩譲って const が再代入不可能 (constant) と言うのはわかるとして、**なぜ let は再代入可能な変数宣言のキーワードとして採用された**のだろうか ?

let といえばML系の言語でよくみられるもので、その多くは一時的なエイリアス(再代入不可)として使われるケースが多いように感じる。
プログラミング以外の文脈 (数学や物理など) で、あるものに名前をつけておくようなケースで let と言う単語持ち出すことはしばしばあって、ML系のキーワードがそこからきていることを考えてもしっくりくる。( where がそれの後置 ver にあたる)

> 大学院時代に理論系の論文でletはたくさん書いた記憶（？）
引用元: [uhyoさんのツイート](https://twitter.com/uhyo_/status/1691630590301507858?s=20)

# 起源
果たして、その let と言うキーワードがどのようにして JavaScript に入ってきて、なぜ今のような挙動をとるようになったのか、またなぜ constとはあまり関係なさそうなのに const と対比されるような登場の仕方をしたのか気になるところである。

そもそも、一番最初に let と言うキーワードが登場したのはいつ、どこなのか。
整理しておくと、EcmaScript に取り込まれたのはみなさんご存知 ES2015. (constもここで入りましたよね)
これを見ると、let / const は同じ時期にJavaScriptに入ってきて、今のような対比 (再代入可/不可) と捉えがちである。
しかし、前述にもある通り、そう考えるのはかなり不自然なところがある。

これについて、そもそも const に関しては Java や C++ 由来で元々 EcmaScript では予約語だったと言う話をいただいた。

> ES5以前からC++やJavaっぽいキーワードは予約語になってたんでconstは自動的に決まって、後からletが増えたの何故か
引用元: [zakkiさんのツイート](https://twitter.com/k_matsuzaki/status/1691600403547451847?s=20)

どうやら let はあとから生えたらしい。

確認してみると、確かに 1997 の段階で const は Future Reserved Words と言う扱いを受けている。
https://www.ecma-international.org/wp-content/uploads/ECMA-262_1st_edition_june_1997.pdf (7.4.3 / P 23)

同様に、2011も見てみると、const は Future Reserved Words と言う扱いを受けている。
https://262.ecma-international.org/5.1/#sec-7.6.1.2

ただ、その下をよくみると、

> The following tokens are also considered to be FutureReservedWords when they occur within [strict mode code (see 10.1.1)](https://262.ecma-international.org/5.1/#sec-10.1.1).

とあり、ここには let も載ってた。

しかしこちらの記載は 1997 (1st) にはなかったので、どこで生えたのか追ってみる。
現在、公式から確認できる資料からだと、2009で初めて現れていた。
https://www.ecma-international.org/wp-content/uploads/ECMA-262_5th_edition_december_2009.pdf (7.6.1.2 / P 211)
4th edition (2009 と 1999の間)の pdf が存在していなかったので、この間に生えたのかは定かではない。

が、少なくとも 1999 (3rd) には存在していなかったので、let と const の登場時期 (キーワードとしての認定)がズレていることは明らかだろう。

余談だが、 同日に let が識別子であるというツイートをして、こちらも備忘録を書いている (https://github.com/Ubugeeei/work-log/discussions/135)
推測するに、let を予約語として取り上げなかったことを考慮して、ユーザー定義の識別子として使っていたものの後方互換性担保の策なのだろう。

同スレに mizchi さんから有力な情報を頂いた。

> うろ覚えなんですが、たしかfirefox でlet が先に生えて、IE（9ぐらい？）でconst 生えてたのがES2015 で標準化した段階でこういう棲み分けになってると記憶してます。たぶん意味論的な棲み分けはなくて、別の出自。
https://twitter.com/mizchi/status/1691641739000545633?s=20

(後で調べると、IE での const 登場時期は IE11 ぽかった)

> 思い出した。最初のfirefox のlet はif 文みたいなブロックがとれたんだった。firefox拡張のコード読んで面食らったことある
https://twitter.com/mizchi/status/1691645433951687083?s=20

こんな構文があったの、全く知らなかった。

https://lia.disi.unibo.it/materiale/JS/developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let.html

このページの Browser compatibility を見てもわかる通り、firefoxが初めて let を導入したのだろうと思われる。

![スクリーンショット 2023-08-17 2 29 18](https://github.com/Ubugeeei/work-log/assets/71201308/007a49f8-fc39-442d-85d0-f527745aa524)

ここに記載されてある、firefox の 2.0 のリリースが 2006年なので、ちょうど 4th edition でゴタゴタしていた時期だとも推測される。
https://en.wikipedia.org/wiki/Firefox_2

一方で、constのサポートがいつからなのかと言うと、
IE 11 (2013)、Chrome 20 (2012)、Firefox 13 (2012)
と言った感じで、リリース日だけ見ると Firefox が先行していたっぽい (?) (実際にどちらが先に実装し始めたのかはわからない)
https://lia.disi.unibo.it/materiale/JS/developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const.html
https://betawiki.net/wiki/Internet_Explorer_11#:~:text=Internet%20Explorer%2011%20is%20the,2008%20R2%20Service%20Pack%201
https://google.fandom.com/wiki/Chrome_version_history

これらを見ると、EcmaScriptとしてのキーワード認定は const の方がだいぶ先だったが、~~実装自体は実は Firefox の let が const とよりもかなり早い段階で行われていたことがわかる。~~
【追記】Mozilla 0.6 ではすでにconstが実装されていたというお話を聞き、アドバイスの通り、ソースコードを読んでみると確かに実装されてありました。つまりは let より const の方が実装としては先です。ログ -> #137 

おそらく、言語の使用を見てみても、firefox は var よりも表現力が豊かで、スタンダードになることを目指して新たな変数宣言 let を実装したんじゃないかと推測できる。~~(この頃、constなんてものはまだ全く登場していない。)~~

Firefox は Mozilla が開発していることでお馴染みだが、Rustの実装を行っていることでもお馴染みである。
let は Rust で変数を宣言するキーワードとしても採用されていて、単純な ident だけでなく、構造を持った式をを引数に受けることができる。そして、Rust は let の他にも match 式や HM 型推論など、ML 系の言語の影響を多く受けていて、Mozilla がそのあたりの言語への関心も高かったと推測できる (ソース不詳。知ってる方いたらぜひ教えてください。)
そう考えると、Firefox で JavaScript に let と言うキーワードが持ち出されたのもなんとなくわかる。

そして、元々は「再代入可能」などといった文脈ではなく、varよりも表現力の高い新たなスタンダードとなるような変数宣言を目指していたのではないかとも推測できる。

# 分かっていない点とか思ったこととか
色々調べてみたが、JavaScript での const の出所が結局わからなかった。
(出所はFirefoxっぽいが、プロパティは守られないと言う点を含めどう言うモチベだったのか)
そして、当初実装されていた let は何がどう言う流れで今の形になったのか、と言うゴタゴタもよくわからなかった。
知ってる方いただぜひコメントください.

色々考えていた時に、let と const はそれぞれのブラウザで違った形で実装されたものをEcmaScriptでねじ込んだがために変な対応関係になっている、みたいなふうに思っていたが、実は出所はどちらも Firefox ? と言うところに行き着いた。
とはいえ、登場時期にかなりのラグがあるので、当初 let と言うキーワードでスタンダードを目指していたであろうものが、「再代入」と言う制限が欲しくなり const が必要になってしまったと言うことはなんとなく考えられる。(JSのようなオブジェクトで参照を持つようなパラダイムではオブジェクト自体が別のものになってしまうと不都合だから? などと勝手に考えている。)

時間が経ち、これらの反省点を元に Rust の変数宣言キーワードが let に集約され、mutability に関しては別のキーワードが当てられるように整理されたとするなら、かなり感慨深いところがある。(ソース不詳)

### shcemeインスパイア
JavaScriptの設計がScheme (Lisp) を意識していると言う意見もあった。 (知らなかった)

> こんにちは。この間たまたま調べた内容なのですが、JavaScriptの設計にあたってブレンダン・アイクさんはScheme (Lisp) を意識していたようです。Schemeにもletがあるので、このアイディアをJSにも持ち込むということを考えた可能性はありそうです
引用元: [SuitCaseさんのツイート](https://twitter.com/pickled_chair/status/1691656896741335072?s=20)

scheme に let と言うキーワードがあるので、それ由来の可能性も !? と言うコメントをいただいた
それにしてはletの登場が遅すぎないか? とは思うものの、遅かったにしろ、let と言うキーワードの選定に関わってきた可能性は大いにある。(調べてみると、letの登場時期 (2006) にブレンダン・アイクさんはMozillaで最高技術責任者をされていたらしい https://ja.wikipedia.org/wiki/%E3%83%96%E3%83%AC%E3%83%B3%E3%83%80%E3%83%B3%E3%83%BB%E3%82%A2%E3%82%A4%E3%82%AF)

**Comment by azu** [comment]
JavaScriptの歴史調査に役立つものをいくつか書いておきます

- [ECMAScript Archives](https://archives.ecma-international.org/)
  - ECMAScript 1 から 2015 策定までのTC39のミーティングログがあります
- [ES Discuss](https://esdiscuss.org/)
  - ECMAScriptのメーリングリストで、ES2015まではよく使われていました
- [archives | Archival material from old web properties.](https://tc39.es/archives/)
  - ES2015(ES6)策定時はBugzillaを使ってタスク管理されていたのでそのアーカイブがあります
- [start [ES Wiki]](https://web.archive.org/web/20151002043637/http://wiki.ecmascript.org/doku.php)
  - 今はGitHubで[ECMAScript proposals](https://github.com/tc39/proposals)を管理していますが、ES2015まではWikiで管理されていたので、ここにProposalがまとまっています
- [JavaScript: The First 20 Years](https://wirfs-brock.com/allen/posts/866)
  - http://www.wirfs-brock.com/allen/jshopl.pdf
  - ES2015のEditorであるAllenさんが書いたJavaScriptの歴史の論文が当時の状況を振り返るのに役立ちます
- ES4はTC39の中でも[ワーキンググループを分けて](https://web.archive.org/web/20141214110502/wiki.ecmascript.org/lib/exe/fetch.php?id=es3.1%3Aes3.1&cache=cache&media=proposals:proposal_to_refocus_tc39-tg1.pdf)やっていたので、ES4についてはこの辺にアーカイブがあります
  - [2006 ECMAScript Archives - Ecma International](https://www.ecma-international.org/ecmascript-development-archive/2006-ecmascript-archives/)
  - [web.archive.org/web/20060615221016/http://developer.mozilla.org:80/es4/](https://web.archive.org/web/20060616010059/http://developer.mozilla.org/es4/proposals/proposals.html)

letについて

ES4時代に、letはSpiderMonkeyのJavaScript 1.7で実装されて、ブロックスコープが主な目的だったようです。
letという語彙を使った理由は調べてないですが、better varの中でできたみたいなのでbinding的な意味合いからきてそうな気はします。(ECMAScriptの仕様書でも `1. Let X be Value.` みたいなステップはよく出てくるので、そこからきている可能性もあります)
当時の状況は[JavaScript: The First 20 Years](https://wirfs-brock.com/allen/posts/866)でも他の言語(Python)などとの比較の話として出てきます。

- [JavaScript 1.7 の新機能: Days on the Moon](https://nanto.asablo.jp/blog/2006/08/12/481381)
- [New in JavaScript 1.7 - JavaScript | MDN](https://web.archive.org/web/20130823145610/https://developer.mozilla.org/en-US/docs/Web/JavaScript/New_in_JavaScript/1.7#Block_scope_with_let_(Merge_into_let_Statement))
- [336378 - (js1.7let) Support let bindings, blocks, and expressions in JS1.7](https://bugzilla.mozilla.org/show_bug.cgi?id=336378)
- ProposalはDave Hermanさん(元Mozilla,Effctive JavaScriptの著者)が書いた[web.archive.org/web/20060616010059/http://developer.mozilla.org/es4/proposals/proposals.html](https://web.archive.org/web/20060613012003/http://developer.mozilla.org/es4/proposals/block_expressions.html)
  - [web.archive.org/web/20060613012003/http://developer.mozilla.org/es4/discussion/block_expressions.html](https://web.archive.org/web/20060615012905/http://developer.mozilla.org/es4/discussion/block_expressions.html)
  - (このProposalを書いたDive Hermanさんの[ブログ](https://web.archive.org/web/20060627045308/http://calculist.blogspot.com/)を見てるとMLの話が多いので、MLに影響受けたのは否定できないかもしれないですね)
  - (このProposalのdiscussionで`let const`という話も出てきています)

ES2015の策定のときには、letとconstはセットにblock scopeとして出ていたようです。

- [harmony:block_scoped_bindings [ES Wiki]](https://web.archive.org/web/20150905155339/http://wiki.ecmascript.org/doku.php?id=harmony:block_scoped_bindings)

語源については当時の周辺状況や予約語などからくるケースは多いので、その辺を調べてみると面白いかもしれません。

**Reply by ubugeeei** [reply]
ありがたすぎます。本当にありがとうございます。
今ゆっくり見れないので仕事終わったらじっくり確認してみます。

**Comment by nanto** [comment]
`const` 自体は Firefox 誕生以前、Netscape 6 (JavaScript 1.5 / Mozilla 0.6 / 2000 年) の時点で実装されていそうです。ただ、その時点ではブロックスコープではなく (`var` と同じく変数の巻き上げが起こる)、また `const` で定義した変数に再代入してもエラーが起きない (変数の値は変わらないので、結果として何も起こらない) という挙動 (変数のプロパティ属性に、ECMA-262 3rd でいうところの ReadOnly 属性を付与するだけ) だったと思われます。(当時の挙動は、[Mozilla 0.6 のソースコード](https://www-archive.mozilla.org/releases/old-releases)の `mozilla/js/` ディレクトリ内を、`JS_HAS_CONST` キーワードを起点に探っていくと推し量れます。)

Firefox 13 というのは再代入でエラーが起きるようになったバージョンのようですね。

**Reply by ubugeeei** [reply]
なるほど〜〜〜〜〜〜〜
これまた全く初見でした。推察の方法までご丁寧にありがとうございます。
developer.mozilla.orgに書かれている該当機能の登場時期とキーワード登場時期がズレてるので、他のもの(const以外)も合わせてソースコードベースでも見てみようかと思います。
キーワードが同じで機能的には祖先には当たるが、前身すぎて同じところにドキュメンテーションされてない、もしくは機能的に祖先ですらないとかありそうなので、ソースコードを見てみる方が僕の知りたいところに辿り着くのは正確そうです。ありがとうございます。

**Comment by ubugeeei** [comment]
Mozilla 0.6 を読むログ -> https://github.com/Ubugeeei/work-log/discussions/137

**Comment by ubugeeei** [comment]
Mozillaの色んなバージョンのソースコードを読むログ -> https://github.com/Ubugeeei/work-log/discussions/138
